File: MAX31855/STM32CubeIDE Example/Core/Src/MAX31855.c
/*************************************************************************************
 Title	 :  MAXIM Integrated MAX31855 Library for STM32 Using HAL Libraries
 Author  :  Bardia Alikhan Afshar <bardia.a.afshar@gmail.com>
 Software:  STM32CubeIDE
 Hardware:  Any STM32 device
*************************************************************************************/
#include"MAX31855.h"
SPI_HandleTypeDef hspi1;

// ------------------- Variables ----------------

uint8_t Error=0;                                      // Thermocouple Connection acknowledge Flag
uint32_t sign=0;									  // Sign bit
uint8_t DATARX[4];                                    // Raw Data from MAX6675
// ------------------- Functions ----------------
float Max31855_Read_Temp(void){
int Temp=0;                                           // Temperature Variable
HAL_GPIO_WritePin(SSPORT,SSPIN,GPIO_PIN_RESET);       // Low State for SPI Communication
HAL_SPI_Receive(&hspi1,DATARX,4,1000);                // DATA Transfer
HAL_GPIO_WritePin(SSPORT,SSPIN,GPIO_PIN_SET);         // High State for SPI Communication
Error=DATARX[3]&0x07;								  // Error Detection
sign=(DATARX[0]&(0x80))>>7;							  // Sign Bit calculation

if(DATARX[3] & 0x07)								  // Returns Error Number
return(-1*(DATARX[3] & 0x07));

else if(sign==1){									  // Negative Temperature
Temp = (DATARX[0] << 6) | (DATARX[1] >> 2);
Temp&=0b01111111111111;
Temp^=0b01111111111111;
return((double)-Temp/4);
	}

else												  // Positive Temperature
{
		Temp = (DATARX[0] << 6) | (DATARX[1] >> 2);
		return((double)Temp / 4);
}
}

File: MAX31855/MicroPython Library/max31855.py
######################################################################################
#Title   :  MAXIM Integrated MAX31855 Library for Raspberry pi Pico
#Author  :  Bardia Alikhan Afshar <bardia.a.afshar@gmail.com>
#Language:  Python
#Hardware:  Raspberry pi pico
#####################################################################################
import ustruct
from machine import Pin, SPI


class MAX31855:
    def __init__(self,SPInum=0,CS=1,MISO=4,SCK=6):
        self.spi = SPI(SPInum, 100_000, polarity=0, phase=0,miso=Pin(MISO),sck=Pin(SCK))
        self.cs =  Pin(CS, Pin.OUT)    #Chip Select Pin
        self.data = bytearray(4)

    def ReadTemp(self):
        self.cs.value(0)             # ChipSelect -> Low to Start SPI
        self.spi.readinto(self.data) # Data Read
        self.cs.value(1)             # ChipSelect -> High to Stop SPI
        error=self.data[3]&0x07      #Thermocouple Error Detection
        if(error):                   # Returns error number and Temp=0              
            Temp=0
            return error, Temp
        sign=(self.data[0]&(0x80))>>7  # Detecting positive or negative temperature
        if sign==1: #Calculation for negative temperature
            Temp = (self.data[0] << 6) | (self.data[1] >> 2)
            Temp&=0b01111111111111
            Temp^=0b01111111111111
            Temp=-Temp*0.25
        else: #Calculation for positive temperature
            Temp = (self.data[0] << 6) | (self.data[1] >> 2)
            Temp=Temp *0.25
        return error, Temp        
  

